src/process_utility.c

```diff
@@ -472,6 +474,225 @@ process_drop_trigger_start(ProcessUtilityArgs *args, DropStmt *stmt)
 	ts_cache_release(hcache);
 }
 
+static DDLResult
+process_drop_schema_start(DropStmt *stmt)
+{
+	/*
+	 * An error will be raised when we start dropping the functions used by a
+	 * background worker, so there is no point in doing anything here.
+	 */
+	if (stmt->behavior == DROP_RESTRICT)
+		return DDL_CONTINUE;
+
+	/*
+	 * Here we are relying on that if we fail to drop one of the
+	 * procedures/functions, this transaction will be rolled back so these
+	 * changes will not be committed.
+	 */
+	ScanIterator iterator =
+		ts_scan_iterator_create(BGW_JOB, RowExclusiveLock, CurrentMemoryContext);
+	ts_scanner_foreach(&iterator)
+	{
+		ListCell *cell;
+		TupleInfo *ti = ts_scan_iterator_tuple_info(&iterator);
+		bool schema_isnull, job_id_isnull;
+		int32 job_id = DatumGetInt32(slot_getattr(ti->slot, Anum_bgw_job_id, &job_id_isnull));
+		Name proc_schema =
+			DatumGetName(slot_getattr(ti->slot, Anum_bgw_job_proc_schema, &schema_isnull));
+		Ensure(!job_id_isnull, "corrupt job entry: job id is null");
+		Ensure(!schema_isnull, "corrupt job entry: schema for job %d is null", job_id);
+		foreach (cell, stmt->objects)
+		{
+#if PG15_GE
+			String *object = lfirst(cell);
+#else
+			Value *object = lfirst(cell);
+#endif
+			if (namestrcmp(proc_schema, strVal(object)) == 0)
+			{
+				CatalogSecurityContext sec_ctx;
+				Assert(stmt->behavior == DROP_CASCADE);
+				ts_catalog_database_info_become_owner(ts_catalog_database_info_get(), &sec_ctx);
+				ereport(NOTICE, errmsg("drop cascades to job %d", job_id));
+				ts_catalog_delete_tid(ti->scanrel, ts_scanner_get_tuple_tid(ti));
+				ts_catalog_restore_user(&sec_ctx);
+			}
+		}
+	}
+	return DDL_CONTINUE;
+}
+
+/*
+ * Start of dropping a procedure.
+ *
+ * We can abort the drop here by throwing an error.
+ */
+static void
+process_drop_procedure_start(DropStmt *stmt)
+{
+	ScanIterator iterator =
+		ts_scan_iterator_create(BGW_JOB, RowExclusiveLock, CurrentMemoryContext);
```

Comment:
Wondering if some of this code about dropping/updating bgw jobs should not be in the bgw code rather then process utility?

Reply:
There is a plethora of different functions doing similar things in `bgw/`. A serious refactoring of the `bgw/` code is overdue and these functions are right now only used here, so better to do a separate refactoring PR and clean up the `bgw` subsystem with a proper set of internal API functions.